src/stream/src/common/log_store/kv_log_store/serde.rs

```diff
@@ -0,0 +1,598 @@
+
+use crate::common::log_store::kv_log_store::{
+    ReaderTruncationOffsetType, RowOpCodeType, SeqIdType,
+};
+use crate::common::log_store::{LogStoreError, LogStoreResult};
+
+/// `epoch`, `seq_id`, `op`
+const PREDEFINED_COLUMNS_TYPES: [DataType; 3] = [DataType::Int64, DataType::Int32, DataType::Int16];
+const EPOCH_INDEX: usize = 0;
+const SEQ_ID_INDEX: usize = 1;
+const ROW_OP_INDEX: usize = 2;
+/// `epoch`, `seq_id`
+const PK_TYPES: [DataType; 2] = [DataType::Int64, DataType::Int32];
+/// epoch
+const EPOCH_TYPES: [DataType; 1] = [DataType::Int64];
+
+const INSERT_OP_CODE: RowOpCodeType = 1;
+const DELETE_OP_CODE: RowOpCodeType = 2;
+const UPDATE_INSERT_OP_CODE: RowOpCodeType = 3;
+const UPDATE_DELETE_OP_CODE: RowOpCodeType = 4;
+const BARRIER_OP_CODE: RowOpCodeType = 5;
+const CHECKPOINT_BARRIER_OP_CODE: RowOpCodeType = 6;
+
+#[derive(Eq, PartialEq, Debug)]
+enum LogStoreRowOp {
+    Row { op: Op, row: OwnedRow },
+    Barrier { is_checkpoint: bool },
+}
+
+#[derive(Clone)]
+pub struct LogStoreRowSerde {
+    /// Used for serializing and deserializing the primary key.
+    pk_serde: OrderedRowSerde,
+
+    /// Row deserializer with value encoding
+    row_serde: BasicSerde,
+
+    /// Serde of epoch
+    epoch_serde: OrderedRowSerde,
+
+    /// Indices of distribution key for computing vnode.
+    /// Note that the index is based on the all columns of the table, instead of the output ones.
+    // FIXME: revisit constructions and usages.
+    dist_key_indices: Vec<usize>,
+
+    /// Virtual nodes that the table is partitioned into.
+    ///
+    /// Only the rows whose vnode of the primary key is in this set will be visible to the
+    /// executor. The table will also check whether the written rows
+    /// conform to this partition.
+    vnodes: Arc<Bitmap>,
+
+    /// The schema of payload
+    payload_schema: Vec<DataType>,
+}
+
+impl LogStoreRowSerde {
+    pub fn new(table_catalog: &Table, vnodes: Option<Arc<Bitmap>>) -> Self {
+        let table_columns: Vec<ColumnDesc> = table_catalog
+            .columns
+            .iter()
+            .map(|col| col.column_desc.as_ref().unwrap().into())
+            .collect();
+        let dist_key_indices: Vec<usize> = table_catalog
+            .distribution_key
+            .iter()
+            .map(|dist_index| *dist_index as usize)
+            .collect();
+
+        let input_value_indices = table_catalog
+            .value_indices
+            .iter()
+            .map(|val| *val as usize)
+            .collect_vec();
+
+        let data_types = input_value_indices
+            .iter()
+            .map(|idx| table_columns[*idx].data_type.clone())
+            .collect_vec();
+
+        // There are 3 predefined columns for kv log store:
+        assert!(data_types.len() > PREDEFINED_COLUMNS_TYPES.len());
+        for i in 0..PREDEFINED_COLUMNS_TYPES.len() {
+            assert_eq!(data_types[i], PREDEFINED_COLUMNS_TYPES[i]);
```

Comment:
Should we check more other than the types to avoid treating regular table column as the prefined column? Like the column name and ordering.

Reply:
Here we only extract the data type, so we can only check the data type. As for avoiding treating regular table column as predefined column, I think we can in future PR add a schema check in the sink executor to ensure that the log store table schema is the same as the sink schema. The payload schema of the log store is the log store table catalog schema without the predefined columns, so with the schema check, we can ensure that we don't treat the regular table column in the sink as the predefined column.